libyul/backends/evm/SSAControlFlowGraphBuilder.cpp

```diff
@@ -626,129 +615,36 @@ SSACFG::ValueId SSAControlFlowGraphBuilder::readVariableRecursive(Scope::Variabl
 	auto& block = m_graph.block(_block);
 	auto& info = blockInfo(_block);
 
-	if (info.sealed && block.entries.size() == 1)
-		return readVariable(_variable, *block.entries.begin());
-
-	SSACFG::ValueId phi = m_graph.newPhi(_block);
-	block.phis.insert(phi);
-	if (info.sealed)
+	SSACFG::ValueId val;
+	if (!info.sealed)
 	{
-		writeVariable(_variable, _block, phi);
-		phi = addPhiOperands(_variable, phi);
+		// incomplete block
+		val = m_graph.newPhi(_block);
+		block.phis.insert(val);
+		info.incompletePhis.emplace_back(val, std::ref(_variable));
 	}
+	else if (block.entries.size() == 1)
+		// one predecessor: no phi needed
+		val = readVariable(_variable, *block.entries.begin());
 	else
 	{
-		info.incompletePhis.emplace_back(phi, std::ref(_variable));
+		// Break potential cycles with operandless phi
+		val = m_graph.newPhi(_block);
+		block.phis.insert(val);
+		writeVariable(_variable, _block, val);
+		val = addPhiOperands(_variable, val);
```

Comment:
Not that it matters that much - and it was me who suggested the flag in ``addPhiOperands`` in the first place :-) - but looking at it again: this is the only other use of ``addPhiOperands``, right? So we could also just *never* try to remove trivial phis in ``addPhiOperands`` and instead just change this here to:
```suggestion
		val = tryRemoveTrivialPhi(addPhiOperands(_variable, val));
```
Might be a bit cleaner...

Reply:
Might also be good to leave a comment at some appropriate place explaining the bug in the paper, in case anyone checks this against the paper again in the future. Wasn't there a link to the paper in the code somewhere as well :-)? I don't see it anymore, but would be good to have it linked somewhere in any case.

Reply:
Huh, that must have gotten lost somewhere, sometime :) I've added comments and a reference to the header